### DISCUSSION TITLE
How does icu calculate phase unwrapping?

### DISCUSSION THREAD

**Discussion by elizabeth-ashley** [isce-framework/isce2#447] [Q&A]
From my understanding, the “Integrated Correlation and Unwrapping” (ICU) method is a modified version of the traditional residue-based unwrapper [Goldstein et al., [1988](https://agupubs.onlinelibrary.wiley.com/doi/pdf/10.1029/RS023i004p00713)] that uses phase gradients, amplitude values, and coherence masking to guide unwrapping using multiple starting seed pixels. 

But this is all I have been able to find on this and would like to understand the math behind the icu unwrapping method. 

Any references or advice is appreciated!

**Comment by gmgunter** [comment]
Unfortunately, I don't believe a detailed description of the ICU algorithm was ever published. I found a description of the algorithm here: [Improved Processing of AIRSAR Data Based on the GeoSAR Processor](https://airsar.jpl.nasa.gov/documents/workshop2002/papers/T3.pdf), but it's pretty high-level (see &#167; 3.5 "Unwrapping and Absolute Phase Determination"). 

I believe the main contributions of ICU, compared to the traditional Goldstein-Zebker-Werner approach, are (A) the use of "neutrons" to guide the creation of branch cuts, (B) application of a low-correlation mask to exclude noisy regions, and (C) the use of tiling & bootstrapping to unwrap arbitrarily long interferograms.

If you have specific questions about the algorithm, I'll do my best to answer them. The [implementation of ICU in isce3](https://github.com/isce-framework/isce3/blob/develop/cxx/isce3/unwrap/icu/Unwrap.cpp) is a little bit cleaner than in isce2, if you want to look at code.

**Reply by elizabeth-ashley** [reply]
@gmgunter this was incredibly helpful! I really appreciate the references and explanation. Here are my notes from what I've learned:
[Unwrapping 101.pdf](https://github.com/isce-framework/isce2/files/8220081/Unwrapping.101.pdf)

I am glad to FINALLY have some clarity on this!  Resources on the topic have been hard to find, like you said seems like the ICU algorithm was not ever published.

I am now trying to understand SNAPHU_MCF and how it compares to ICU. I've read the Chen and Zebker 2001 paper, but if you have any resources on this topic I'd welcome any advice you can give me! Thank you!

**Reply by gmgunter** [reply]
>Here are my notes from what I've learned:
>Unwrapping 101.pdf

Looks good, nice work.

I would make a distinction between "absolute phase" and "unwrapped phase". The AIRSAR paper above doesn't really do much to differentiate phase unwrapping from absolute phase determination, but in general they could be distinct steps.

*Absolute phase* is the unambiguous phase proportional to the differential range between the interferometric pair

<p align="center">
<img src="https://latex.codecogs.com/svg.image?\Phi_{abs}&space;=&space;4\pi&space;\delta_r&space;/&space;\lambda">
</p>

   where <img src="https://latex.codecogs.com/svg.image?\delta_r"> is the one-way path length difference and <img src="https://latex.codecogs.com/svg.image?\lambda"> is the radar wavelength.

*Unwrapped phase* is the result of phase unwrapping, which attempts to correct for the relative difference in phase between pixels within the interferogram. It differs from the absolute phase by a fixed integer multiple of 2pi

<p align="center">
<img src="https://latex.codecogs.com/svg.image?\Phi_{unw}&space;=&space;\Phi_{abs}&space;&plus;&space;2\pi&space;n">
</p>

Absolute phase determination is the process of trying to estimate the value of <img src="https://latex.codecogs.com/svg.image?n"> (possibly by minimizing the height ambiguity against an existing, low resolution DEM). The implementation of ICU in isce3 does not perform this step -- I suspect it's the same in isce2.

If we're measuring topographic height, it's necessary to estimate the absolute phase, but if we're trying to measure deformation signals, we probably don't care about the absolute phase and can stop after the unwrapping step.

>I am now trying to understand SNAPHU_MCF and how it compares to ICU. I've read the Chen and Zebker 2001 paper, but if you have any resources on this topic I'd welcome any advice you can give me!

Sure. [[Chen, Zebker 2000]](https://opg.optica.org/josaa/fulltext.cfm?uri=josaa-17-3-401&id=59155) is a precursor to the 2001 paper you mentioned and presents a nice review of network-based phase unwrapping. I would also check out [[Costantini 1998]](https://ieeexplore.ieee.org/document/673674), which AFAIK is the original paper to introduce a network-optimization approach for 2DPU. It does a good job of presenting a mathematical framework for the network-based unwrapping approaches.

The MCF approach seeks to minimize the total (weighted) number of extra gradient cycles that must be added to the wrapped phase in order to unwrap it. It turns out that this is exactly equivalent to a [Minimum Cost Flow problem](https://en.wikipedia.org/wiki/Minimum-cost_flow_problem) and there are efficient algorithms for computing this. 

The idea is that this is a *global* optimization approach that is able to make use of information from the full interferogram when deciding where to insert extra phase cycles. By contrast, the branch-cut method uses a greedy algorithm that is susceptible to falling into local optima which may correspond to large unwrapping errors.

SNAPHU similarly adopts methods from the field of network optimization but differs from the MCF approach in its choice of optimization objective. It cleverly formulates the phase unwrapping problem as a [MAP estimation](https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation) problem using statistical models for the unwrapped phase gradients.

The downside of SNAPHU is that its cost surface is non-convex, whereas the MCF problem is convex. As a result, SNAPHU must resort to much less efficient solver algorithms, and is not guaranteed to converge to a globally optimal solution.

**Comment by jason2318** [comment]
i am reading your notes!very cute lass.love from China.030
